src/crypto/CHIPCryptoPAL.cpp

```diff
@@ -159,10 +162,70 @@ CHIP_ERROR ConvertIntegerRawToDerInternal(const ByteSpan & raw_integer, MutableB
     return CHIP_NO_ERROR;
 }
 
-} // namespace
+/**
+ * @brief Find a 4 uppercase hex digit hex value after a prefix string. Used to implement
+ *        fallback CN VID/PID encoding for PAA/PAI/DAC.
+ *
+ * @param[in] buffer - buffer in which to find the substring.
+ * @param[in] prefix - prefix to match, which must be followed by 4 uppercase hex characters
+ * @param[out] out_hex_value - on CHIP_NO_ERROR return, this will be the 16-bit hex value decoded.
+ * @return CHIP_NO_ERROR on success, CHIP_ERROR_NOT_FOUND if not detected and
+ *         CHIP_ERROR_WRONG_CERT_DN if we saw the prefix but no valid hex string.
+ */
+CHIP_ERROR Find16BitUpperCaseHexAfterPrefix(const ByteSpan & buffer, const char * prefix, uint16_t & out_hex_value)
+{
+    chip::CharSpan prefix_span = chip::CharSpan::fromCharString(prefix);
 
-namespace chip {
-namespace Crypto {
+    bool found_prefix_at_least_once = false;
+
+    // Scan string from left to right, to find the desired full matching substring.
+    for (size_t start_idx = 0; start_idx < buffer.size(); start_idx++)
```

Comment:
We need prefix_span.size() + HEX_ENCODED_LENGTH(sizeof(uint16_t)), so I wonder if we could make the check and the loop in one and never worry about size checks after that again. Something like:

```cpp
const size_t data_len = prefix_span.size() + HEX_ENCODED_LENGTH(sizeof(uint16_t));

if (buffer.size() < data_len) 
{
  return CHIP_ERROR_NOT_FOUND; // insufficient space
}

for (const uint8_t cursor = buffer.data(); cursor < buffer.data() + buffer.size() - data_len; cursor++) 
{
   if (memcmp(cursor, prefix_span.data(), prefix_span.size()) != 0)
   {
      continue;
   }

   found_prefix_at_least_once = true;

   if (Encoding::UppercaseHexToUint16(
              reinterpret_cast<const char *>(cursor + prefix_span.size()),
              HEX_ENCODED_LENGTH(sizeof(uint16_t)),
              out_hex_value) != 0)
   {
       // Found first full valid match, return success, out_hex_value already updated.
       return CHIP_NO_ERROR;
   }
}
// ...
```

Reply:
I also wonder if we should cast the `buffer` to char* to begin with and maybe even use strstr to find prefix occurences.

Reply:
I guess buffer can contain null terminators in it so strstr does not work.

Reply:
Exactly: no strstr allowed.